packages/daimo-api/src/server/router.ts

```diff
@@ -624,6 +635,90 @@ export function createRouter(
         return status.isValid;
       }),
 
+    onAccountUpdate: publicProcedure
+      .input(
+        z.object({
+          address: zAddress,
+          inviteCode: z.string().optional(),
+          sinceBlockNum: z.number(),
+        })
+      )
+      .subscription(async (opts) => {
+        const { address, inviteCode } = opts.input;
+        // how often to send updates regardless of new transfers
+        // useful to update exchange rates and others.
+        const refreshInterval = 10_000;
+
+        return observable<AccountHistoryResult>((emit) => {
+          let lastEmittedBlock = opts.input.sinceBlockNum;
+          let getAccountHistoryPromise: Promise<AccountHistoryResult> | null =
+            null;
+
+          const pushHistory = (onlyOnNewTransfers: boolean) => {
+            getAccountHistoryPromise = getAccountHistory(
+              opts.ctx,
+              address,
+              inviteCode,
+              lastEmittedBlock,
+              watcher,
+              vc,
+              homeCoinIndexer,
+              ethIndexer,
+              foreignCoinIndexer,
+              profileCache,
+              noteIndexer,
+              reqIndexer,
+              inviteCodeTracker,
+              inviteGraph,
+              nameReg,
+              keyReg,
+              paymaster,
+              db
+            );
+
+            getAccountHistoryPromise
+              .then((history) => {
+                // we can have concurrent requests. discard those that arrived too late
+                if (history.lastBlock < lastEmittedBlock) {
+                  return;
+                }
+
+                if (onlyOnNewTransfers && history.transferLogs.length === 0) {
+                  return;
+                }
+
+                emit.next(history);
+
+                lastEmittedBlock = history.lastBlock;
+              })
+              .finally(() => {
+                getAccountHistoryPromise = null;
+              });
+          };
+
+          const eventListener = () => {
+            pushHistory(true);
+          };
+
+          const intervalTimer = setInterval(() => {
+            // interval concided with new block. let's skip this one.
+            if (getAccountHistoryPromise) {
+              return;
+            }
+
+            pushHistory(false);
+          }, refreshInterval);
```

Comment:
this is quite tricky... if I'm understanding correctly:

- interval fetch skipped if we're currently doing an event-based fetch
- however, event-based fetch is *not* skipped if we're doing an interval fetch

in that second case, both can run concurrently.

Reply:
Yes, your interpretation is correct. In both cases we favor event-triggered sync over interval one because it is guaranteed to have latest data.
For 1) we skip interval because data was requested before Shovel committed data to Postgres. For 2) we do don't do interval-triggered query because there is already event-triggered query running, which will have the latest data.

Reply:
Aha! I just noticed an unexpected edge case for case 2: when interval sync coincides with event-triggered sync and there is no new transfer for that block, user will have to wait for another interval sync to get updated about other `AccountHistoryResult` changes (namely keys, exchange rates, linked accounts, named accounts, suggested actions, invites, notifications requests, and proposed swaps.)

Thanks for @dcposch!